lib/http.c

```diff
@@ -980,6 +983,169 @@ static int is_valid_auth_separator(char ch)
 }
 #endif
 
+struct authdata {
+  const char *auth;
+  struct Curl_easy *data;
+  bool proxy;
+  struct connectdata *conn;
+  curlnegotiate *negstate;
+  unsigned long *availp;
+  struct auth *authp;
+};
+
+static bool handle_spnego(struct authdata *authinfo)
+{
+  (void)authinfo;
+#ifdef USE_SPNEGO
+  if(checkauth("Negotiate", authinfo->auth)) {
+    if((authinfo->authp->avail & CURLAUTH_NEGOTIATE) ||
+       Curl_auth_is_spnego_supported()) {
+      *authinfo->availp |= CURLAUTH_NEGOTIATE;
+      authinfo->authp->avail |= CURLAUTH_NEGOTIATE;
+
+      if(authinfo->authp->picked == CURLAUTH_NEGOTIATE) {
+        CURLcode result = Curl_input_negotiate(authinfo->data,
+                                               authinfo->conn,
+                                               authinfo->proxy,
+                                               authinfo->auth);
+        if(!result) {
+          free(authinfo->data->req.newurl);
+          authinfo->data->req.newurl = strdup(authinfo->data->state.url);
+          if(!authinfo->data->req.newurl)
+            return CURLE_OUT_OF_MEMORY;
```

Comment:
This returns an error code, not a boolean! You probably need to propagate the error code.